doc/design/proposal-synchronous-hooks.md

```diff
@@ -0,0 +1,334 @@
+# Universal, synchronous and in-thread loader hooks
+
+For background and motivation, see https://github.com/nodejs/node/issues/52219. Prototype implementation is in https://github.com/joyeecheung/node/tree/sync-hooks.
+
+TL;DR: the top priority of this proposal is to allow sun-setting CJS loader monkey patching as non-breakingly as possible (which is why it needs to be synchronous and in-thread because that's how the CJS loader and `require()` works). Then it's API consistency with the existing `module.register()` off-thread hooks.
+
+Existing users of CJS loader monkey patching to look into:
+
+- pirates (used by Babel and nyc)
+- require-in-the-middle (used by many tracing agents)
+- Yarn PnP
+- tsx
+- ts-node
+- proxyquire, quibble, mockery
+
+## API design
+
+A high-level overview of the API.
+
+```js
+const { registerHooks } = require('module');
+
+function resolve(specifier, context, nextResolve) {
+  const resolved = nextResolve(specifier, context);
+  resolved.url = resolved.url.replaceAll(/foo/g, 'bar');
+  return resolved;
+}
+
+function load(url, context, nextLoad) {
+  const loaded = nextLoad(specifier, context);
+  loaded.source = loaded.source.toString().replaceAll(/foo/g, 'bar');
+  return loaded;
+}
+
+const hook = registerHooks({ resolve, load });
```

Comment:
Rebased a bit and I noticed that if this takes an specifier/URL as @guybedford requested the resolution be quite weird, because previously `module.register(specifier, parentURL)` assumes that the resolution takes the `import` condition, which is going to be awkward for the universal hooks that support both `require` and `import` and can be loaded either way and authored in either module format for backward compat. I guess for now the best we can do is to suggest users to use neither `import` nor `require` for the resolution of the file to be registered (conceptually it makes no sense for either to apply, because this is not loaded with a new mechanism). But this still leaves a bad taste in my mouth. 

It may also be somewhat awkward when being combined with https://github.com/nodejs/node/issues/53787 where the preload is added in an external API/config, then users need two files, one for the preload and one for the hooks, which is less nice than just preloading one file for them all. They would also need to disable the inheritance that comes from this hook-specific setting because the preload config may already be able to specify a more flexible inheritance and you don't want it to get preloaded twice.